nexus/src/sagas.rs

```diff
@@ -475,9 +488,77 @@ async fn sic_create_network_interfaces_undo(
     Ok(())
 }
 
+async fn sic_attach_disks_to_instance(
+    sagactx: ActionContext<SagaInstanceCreate>,
+) -> Result<(), ActionError> {
+    ensure_instance_disk_attach_state(sagactx, true).await
+}
+
+async fn sic_attach_disks_to_instance_undo(
+    sagactx: ActionContext<SagaInstanceCreate>,
+) -> Result<(), anyhow::Error> {
+    Ok(ensure_instance_disk_attach_state(sagactx, false).await?)
+}
+
+async fn ensure_instance_disk_attach_state(
+    sagactx: ActionContext<SagaInstanceCreate>,
+    attached: bool,
+) -> Result<(), ActionError> {
+    let osagactx = sagactx.user_data();
+    let saga_params = sagactx.saga_params();
+    let opctx =
+        OpContext::for_saga_action(&sagactx, &saga_params.serialized_authn);
+
+    let saga_disks = &saga_params.create_params.disks;
+    let instance_name = sagactx.lookup::<db::model::Name>("instance_name")?;
+
+    let organization_name: db::model::Name =
+        saga_params.organization_name.clone().into();
+    let project_name: db::model::Name = saga_params.project_name.clone().into();
+
+    for disk in saga_disks {
```

Comment:
(Not explicitly a request, more just out-loud thinking)

To clarify the path, from a high-level:

- `sic_attach_disks_to_instance` happens right before actually sending the "instance ensure" request
- It calls `Nexus::instance_attach_disk`, which calls `Nexus::disk_set_runtime`
- If the instance state is `InstanceState::Creating` and the requested disk state is `DiskStateRequested::Attached`, we write the requested disk state ("attached") to the database, but don't send any requests to the sled agent (yet).

I wonder if we want to bifurcate `disk_set_runtime` into two different functions, within Nexus. The fact that it "may or may not make a request to sled agent" feels a little unintuitive to me - I 100% understand why it's acting that way, and I think deferring the request until actually calling "instance ensure" makes sense, but understanding when it would / would not make a call to sled agent feels a little non-obvious.

WDYT? I think this is subjective, but just wanna make sure the flows are as straightforward as they can be.

Reply:
This is true (though the actual path traversed through the code has changed as part of 581bdeef1926b6e6af0c76fb32cb510f62a415f0), with the idea that attaching disks to the instance only in the DB is preparing for the logic in instance_set_runtime that queries the disks attached to the instance.

We may want to remove the `disk_put` path and do everything through `instance_put`. That way we can funnel changes through `instance_set_runtime` and let propolis handle the disk attach or remove as it needs to.